versions/development/SPEC.md

```diff
@@ -1218,81 +1270,316 @@ $runtime = 'runtime' $ws* '{' ($ws* $runtime_kv $ws*)* '}'
+
+During execution of a task, resource requirements within the runtime section must be enforced by the engine. If the engine is not able to provision the requested resources, then the task should immediately fail prior to the start of execution. In order to ensure adequate information is provided to execute a task, there are a number or required keys, which are defined below.
+
+Values can be any expression which evaluate to a valid entry for that runtime property. Since values are expressions, they can also reference variables in the task:
 
 ```wdl
 task test {
+  input {
+    String ubuntu_version
+  }
   command {
     python script.py
   }
+  
   runtime {
-    docker: ["ubuntu:latest", "broadinstitute/scala-baseimage"]
+    container: ubuntu_version
+    cpu: 1
+    memory: "1 GB"
   }
 }
 ```
 
-The value for the `docker` runtime attribute in this case is an array of values.  The parser should accept this.  Some engines might interpret it as an "either this image or that image" or could reject it outright.
 
-Since values are expressions, they can also reference variables in the task:
+#### container (**Required**)
 
+The container key must accept one ore more locations which inform the engine where to retrieve a container image to execute the task. It is expected (but not enforced) that each container image provided are identical, and will provide the same final results when the task is run. It is the responsibility of the individual execution engine to define the specific image sources which it supports, and to determine which image is the "best" one to use at runtime. Defining multiple images enables greater portability across a broad range of execution environments. 
+
+Container source locations should use the syntax defined by the individual container repository. For example an image defined as `ubuntu:latest` would conventionally refer a docker image living on `DockerHub`, while an image defined as `quay.io/bitnami/python` would refer to a `quay.io` repository. 
+
+The supported attributes are:
+
+* `String`: A single container location
+* `Array[String]`: A list of container entries. **Note**: the ordering of the list does not imply any implicit preference or ordering of the containers. All images are expected to be the same, and therefore any choice would be equally valid.
+
+**Single Image**
 ```wdl
-task test {
-  input {
-    String ubuntu_version
-  }
-  command {
-    python script.py
+task single_image_test {
+  #....
+  runtime {
+    container: "ubuntu:latest"
   }
+```
+
+**Multiple Images**
+```wdl
+task multiple_image_test {
+  #.....
   runtime {
-    docker: "ubuntu:" + ubuntu_version
+    container: ["ubuntu:latest", "gcr.io/standard-images/ubuntu:latest"]
   }
 }
 ```
 
-Most key/value pairs are arbitrary.  However, the following keys have recommended conventions:
 
-#### docker
+#### cpu (**Required**)
```

Comment:
I'm nervous about increasing the amount of boilerplate this means adding into every. single. WDL. task. definition.

Can we instead say "if a `cpu` value is not specified then the task must work with exactly '1' CPU"? And something similar for `memory`?

Reply:
Hmm, that is a good suggestion, hardcoding the default in the spec. my worry though, is that people will just ignore that entirely and continue on using things like the `default-runtime` attributes, and relying on things that are very cromwell or other engine specific. This is what lead to alot of the motivation for redefining the runtime block. I suppose we could state:

```
If a `cpu` key is not defined, then then it must be interpreted that the task will run with exactly 1 CPU.  engines should not provide a mechanism to set their own default values which circumvents this
```

Reply:
@cjllanwarne the other idea I had, was to allow for a runtime section in the 'workflow' definition. This would allow for the definition of runtime defaults which are explicitly stated and it Also provides a mechanism for documenting them.

Reply:
@cjllanwarne on second thought, I Think I prefer enforcing an implicit definition of a default within the WDL as opposed to allowing defaults at the `workflow` level. Not every task will have a workflow, therefore people will likely abuse this approach for tasks and just assunme a user will define the default value